webrtc/src/rtp_transceiver/rtp_sender/mod.rs

```diff
@@ -198,7 +176,85 @@ impl RTCRtpSender {
             paused: Arc::new(AtomicBool::new(start_paused)),
 
             internal,
+        };
+
+        sender.add_encoding_internal(track).await;
+
+        // Add track
+        sender
+    }
+
+    pub async fn add_encoding(&self, track: Arc<dyn TrackLocal + Send + Sync>) -> Result<()> {
+        if track.rid() == "" {
+            return Err(Error::ErrRTPSenderRidNil);
+        }
+
+        if self.has_stopped().await {
+            return Err(Error::ErrRTPSenderStopped);
         }
+
+        if self.has_sent().await {
+            return Err(Error::ErrRTPSenderSendAlreadyCalled);
+        }
+
+        // oops, somebody code-golf this for me
+        {
+            let ref_track = if let Some(t) = self.first_encoding().await? {
+                let t = t.track.lock().await;
+                if let Some(t) = &*t {
+                    if t.rid() != "" {
+                        t.clone()
+                    } else {
+                        return Err(Error::ErrRTPSenderNoBaseEncoding);
+                    }
+                } else {
+                    return Err(Error::ErrRTPSenderNoBaseEncoding);
+                }
+            } else {
+                return Err(Error::ErrRTPSenderNoBaseEncoding);
+            };
+
+            if ref_track.id() != track.id()
+                || ref_track.stream_id() != track.stream_id()
+                || ref_track.kind() != track.kind()
+            {
+                return Err(Error::ErrRTPSenderBaseEncodingMismatch);
+            }
+
+            if self.encoding_for_rid(track.rid()).await.is_some() {
+                return Err(Error::ErrRTPSenderRIDCollision);
+            }
+        }
+
+        self.add_encoding_internal(track).await;
+        Ok(())
+    }
+
+    pub(crate) async fn add_encoding_internal(&self, track: Arc<dyn TrackLocal + Send + Sync>) {
+        let ssrc = rand::random::<u32>();
+        let srtp_stream = Arc::new(SrtpWriterFuture {
+            closed: AtomicBool::new(false),
+            ssrc,
+            rtp_sender: Arc::downgrade(&self.internal),
+            rtp_transport: Arc::clone(&self.transport),
+            rtcp_read_stream: Mutex::new(None),
+            rtp_write_session: Mutex::new(None),
+        });
+
+        let srtp_rtcp_reader = Arc::clone(&srtp_stream) as Arc<dyn RTCPReader + Send + Sync>;
```

Comment:
No idea, this is ripped directly from Pion and I hadn't given any thoughts to whether any of it was unnecessary. I'll have a look at this tomorrow when I've pushed my other pending tweaks

Reply:
Okay it looks like I hand-waved when copying this from Pion because the code has diverged quite a bit. 

I guess my preference on first glance would be to not duplicate the infrastructure however the complication/motivation however for doing *some* duplication is that to make simulcast work we need to call bind_local_stream for each outgoing encoding where we inspect that stream, and wrap that writer (and just that writer) with the code that injects the headers for MID/RID so simulcast works.

I'll need to read the code properly to understand what the correct way of achieving this is. I agree that simply cloning an Arc looks to be insufficient for this task.

Reply:
(It clearly works at the moment, so whatever is being cloned is sufficently stateless that I got away with it)

Reply:
Hmm, if you are leaving it like this can we add a `TODO` about this point?